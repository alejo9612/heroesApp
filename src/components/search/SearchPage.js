import React, { useMemo } from 'react';
import queryString  from 'query-string';
//import { heroes } from '../../data/heroes';
import { useForm } from '../../hooks/useForm';
import { HeroeCard } from '../heroes/HeroeCard';
import { useLocation } from 'react-router';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchPage = ({history}) => {

    const location = useLocation();
    //comprobamos lo que nos almacena el paquete de node
    //console.log(queryString.parse(location.search));
    const {q = ''} = queryString.parse(location.search);

    const [FormValues, handleInputChange] = useForm({
        BuscarText: q
    });

    const {BuscarText} = FormValues;

    //const heroesFiltrados = heroes;
    //const heroesFiltrados = useMemo(() => getHeroesByName(BuscarText), [BuscarText]);
    const heroesFiltrados = useMemo(() => getHeroesByName(q), [q]);

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push(`?q=${BuscarText}`)

        // console.log(BuscarText)
    }

    return (
        <div>
            <h1>Buscar Heroe</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form className="d-grid gap-2"
                          onSubmit={handleSubmit}
                    >
                        <input className="form-control"
                               placeholder="Buscar Heroe"
                               type="text"
                               name="BuscarText"
                               autoComplete="off"
                               value={BuscarText}
                               onChange={handleInputChange}
                        />
                        <button className="btn m-1 btn-outline-primary"
                            type="submit"
                        >
                            Buscar..
                    </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr />
                    {
                        (q === '') &&
                        <div className="alert alert-info">
                            Busca un heroe por favor!!
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltrados.length === 0 ) &&
                        <div className="alert alert-danger">
                            El nombre "{q}" no es un superheroe!!
                        </div>
                    }

                    {
                        heroesFiltrados.map(heroe => (
                            <HeroeCard key={heroe.id}
                                {...heroe}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
