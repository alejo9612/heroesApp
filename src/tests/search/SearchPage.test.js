import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchPage } from "../../components/search/SearchPage";

describe('Se llevar√° a cabo las pruebas del <SearchPage/>', () => {

    test('Debe de mostrarse de manera correcta con sus valores por default', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/Buscar']}>
                <Route path="/Buscar" component={SearchPage} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe por favor!!')

    });

    test('Debe de mostrar a batman y el input con valor de query', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/Buscar?q=batman']}>
                <Route path="/Buscar" component={SearchPage} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe de mostrar un error si no se encuetra el Heroe', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/Buscar?q=batman123']}>
                <Route path="/Buscar" component={SearchPage} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').prop('value')).toBe(`El nombre batman123 no es un superheroe!!`);
        expect(wrapper).toMatchSnapshot();

    });
    
    test('Debe de llamar el PUSH del history ', () => {
        
        const history = {
            push:jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/Buscar?q=batman123']}>
                <Route path="/Buscar" component={()=> <SearchPage history={history}/>} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target:{
                name: 'Buscar',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith(`?q=batman`);
        
    });
    
});
