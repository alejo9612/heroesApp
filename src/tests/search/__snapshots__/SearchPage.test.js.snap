// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Se llevará a cabo las pruebas del <SearchPage/> Debe de mostrar a batman y el input con valor de query 1`] = `
<div>
  <h1>
    Buscar Heroe
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="d-grid gap-2"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="BuscarText"
          onChange={[Function]}
          placeholder="Buscar Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-outline-primary"
          type="submit"
        >
          Buscar..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="card ms-3 mt-4"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                className="btn btn-primary"
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Se llevará a cabo las pruebas del <SearchPage/> Debe de mostrarse de manera correcta con sus valores por default 1`] = `
<div>
  <h1>
    Buscar Heroe
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="d-grid gap-2"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="BuscarText"
          onChange={[Function]}
          placeholder="Buscar Heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-outline-primary"
          type="submit"
        >
          Buscar..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busca un heroe por favor!!
      </div>
    </div>
  </div>
</div>
`;
